src/Admin/Controller/UpdateExtensionController.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+/*
+ * This file is part of Flarum.
+ *
+ * For detailed copyright and license information, please view the
+ * LICENSE file that was distributed with this source code.
+ */
+
+namespace Flarum\Admin\Controller;
+
+use Flarum\Extension\ExtensionManager;
+use Flarum\Http\RequestUtil;
+use Flarum\Http\UrlGenerator;
+use Illuminate\Support\Arr;
+use Laminas\Diactoros\Response\RedirectResponse;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface as Request;
+use Psr\Http\Server\RequestHandlerInterface;
+
+class UpdateExtensionController implements RequestHandlerInterface
+{
+    /**
+     * @var UrlGenerator
+     */
+    protected $url;
+
+    /**
+     * @var ExtensionManager
+     */
+    protected $extensions;
+
+    public function __construct(UrlGenerator $url, ExtensionManager $extensions)
+    {
+        $this->url = $url;
+        $this->extensions = $extensions;
+    }
+
+    /**
+     * @throws \Flarum\User\Exception\PermissionDeniedException
+     */
+    public function handle(Request $request): ResponseInterface
+    {
+        $actor = RequestUtil::getActor($request);
+
+        $actor->assertAdmin();
+
+        $enabled = (bool) (int) Arr::get($request->getParsedBody(), 'enabled');
+        $name = Arr::get($request->getQueryParams(), 'name');
+
+        if ($enabled === true) {
+            $this->extensions->enable($name);
+        } elseif ($enabled === false) {
+            $this->extensions->disable($name);
+        }
```

Comment:
we could do that, we could also create a toggle extension command handler and re-use that, similarly to how we re-use other commands in other places, which is a more simple process.

```php
$this->bus->dispatch(
    new ToggleExtension($extensionId, $enabled)
);
```

Reply:
> we could do that, we could also create a toggle extension command handler and re-use that, similarly to how we re-use other commands in other places, which is a more simple process.

IMO the biggest factor for determining between the two is whether we'd like to move to more of an auto-generated, graphql-like solution. In that case, we'd want API endpoints to represent the "source of truth" for this logic.

Reply:
Sorry if I'm missing something, I know very little about graphql.

That said, generally speaking, I prefer using the command bus when for *actions*, right now we use the API client for reading data to include in nojs views. But when it comes to taking actions, we re-use commands, and that makes sense because *commands* do represent the source of truth for the logic, but also because we are directly interacting with it, so we get the proper returned values, whereas using the API client we'd get a JSON response as a return value, and exceptions (*I think right now*) would only be returned within that JSON response and not be thrown. And that seems like introducing more complexity. What happens when an extension dependant on another is enabled/disabled ?

Reply:
> exceptions (I think right now) would only be returned within that JSON response and not be thrown

that is something we should change #3049, but for now that's the behavior we have to deal with yeah.

I agree with you then, a command would probably be best here.